### PR TITLE
fix: audit #70 resilience follow-up (P0/P2/P3/P4)

### PR DESCRIPTION
**Pull request by DIvyaNautiyal07**
## Summary

Closes the 9 findings from issue #70 — the post-#61/#62/#63 audit. Six in-repo fixes plus three resilience improvements pulled in via a hash bump on the upstream genai+x402 connection package.

Depends on **valory-xyz/genai#30** (already opened, hashes published to IPFS so this PR's CI can sync them).

## In-repo fixes

- **P0-A1** `agent_db_client.py` — use `response.body!r` instead of the missing `.text` attribute. The previous f-string raised `AttributeError` on any non-success non-404 response, escaping the `RuntimeError`-only handlers above and crash-looping the agent.
- **P2-2** `handlers.py` — chatui POST handler now wraps `json.loads` in try/except and surfaces malformed bodies as 400 instead of letting the dispatch wrapper catch and return 500.
- **P3-1** `mirror_db/connection.py` — retry decorator now retries on 502/503/504 in addition to 429 and connection errors. 500 stays non-retryable (it's most often a programming error, not a transient hiccup).
- **P4-1** `behaviour_classes/base.py` — meme-subgraph response chain guards against partial-null at every level. The previous chain only guarded the top-level `data: null` case; `data.memeTokens: null` raised `AttributeError` outside the per-token try/except.
- **P4-2** `behaviour_classes/base.py`, `twitter.py` — deleted the two dead-code `get_memeooorr_handles_from_{subgraph,chain}` parsers, their tests, and the imports/constants used only by them (`ServiceRegistryContract`, `MEMEOOORR_DESCRIPTION_PATTERN`, `IPFS_ENDPOINT`, the unused `re` import). The only call site in `twitter.py` was already commented out behind a "Disabled loading agents from subgraph" warning.
- Also removed the now-unused `service_registry_l2` contract dependency from `memeooorr_abci/skill.yaml`.

## Hash bumps

Carrying P0-S2, P2-1, P0-S1 from valory-xyz/genai#30:

- `valory/genai:0.1.0` → `bafybeifyqyvrg3w4hz2uwhxpfngmgbcrcsu4xfp4eutrcyte2h3zicgzbi`
- `valory/x402:0.1.0` → `bafybeiae7y5sdfkdtmrdl4tcgniinpul4tie6rsde6to3y3lxdkorhgac4`

If the upstream PR receives review changes, the released hashes will differ and this PR's `chore(packages): bump genai + x402 connection hashes` commit will need re-rolling against the post-merge hashes.

## Test plan

- [ ] `uv run autonomy packages lock --check` passes
- [ ] `uv run tox -e py3.12-darwin` passes (1349 tests, all pass locally)
- [ ] `uv run tox -p -e black-check -e isort-check -e flake8 -e mypy -e pylint -e darglint` all pass locally
- [ ] `tox -e check-hash`, `check-packages`, `check-handlers`, `check-abci-docstrings`, `check-abciapp-specs`, `check-dependencies`, `check-doc-hashes`, `check-third-party-hashes` all pass locally
- [ ] Local single-agent run starts cleanly with the path-based .env

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by LOCKhart07 — COMMENTED** [review]
Non-blocking — the actual fixes look correct and well-tested. Two cleanup observations:

**1. Incomplete `service_registry_l2` cleanup**

The PR description says the `service_registry_l2` dependency was removed, but only the entry in `memeooorr_abci/skill.yaml` was dropped. Two places still declare it:

- `packages/valory/agents/memeooorr/aea-config.yaml:26` (agent-level contract dep)
- `packages/packages.json:4` (dev section)

After this PR, no Python module under `packages/valory/skills/` imports `ServiceRegistryContract`. Either drop these entries too (and re-lock) or correct the PR description.

**2. Orphaned helpers and params in `behaviour_classes/base.py`**

The dead-code deletion correctly removed the two parser methods, but it leaves behind code that was only reachable from them:

- `get_service_registry_address()` (base.py:652) — only caller was the deleted `get_memeooorr_handles_from_chain`
- `get_packages()` (base.py:603) — only caller was the deleted `get_memeooorr_handles_from_subgraph`; tests still exercise it but no production code does
- Params `service_registry_address_base`, `service_registry_address_celo`, `olas_subgraph_url` declared in `skill.yaml` / `aea-config.yaml` / `service.yaml` / `models.py` — reachable only from the orphaned helpers above

Same rationale that justified deleting the parsers ("if anyone needs to revive this lookup they should write fresh code") applies here. Fine to defer to a follow-up, but worth flagging.

---

What's solid:

- **P0-A1**: `response.body!r` is consistent with the four other `.body` references in `_request`. Test swap to `SimpleNamespace` is a real improvement — guards against re-introducing `.text`.
- **P2-2**: catch is scoped to `(UnicodeDecodeError, json.JSONDecodeError)`, surfaces 400 via `_handle_bad_request`. Tests cover both non-UTF8 and truncated-JSON.
- **P3-1**: 502/503/504 retry path is parametrized; 500-not-retryable, max-retries-reached, and 503-then-success all tested. Rationale for excluding 500 is documented inline.
- **P4-1**: new chain `((data or {}).get('memeTokens') or {}).get('items') or []` is strictly more defensive; tests cover null at every depth.
- **P4-2**: no stale references to `MEMEOOORR_DESCRIPTION_PATTERN`, `IPFS_ENDPOINT`, or the deleted methods anywhere in the source tree.

**Comment by DIvyaNautiyal07** [comment]
Both observations were grounded — addressed in `dc240fdc`. Brief breakdown:

**1. service_registry_l2 cleanup**

Dropped the contract from `agents/memeooorr/aea-config.yaml` and the `dev` section of `packages.json`. Also deleted the contract package directory itself (`packages/valory/contracts/service_registry_l2/`) since nothing in this repo imports `ServiceRegistryContract` any longer. Re-locked, all check-hash / check-packages / check-handlers pass.

**2. Orphaned helpers and params**

Deleted `get_service_registry_address` and `get_packages` from `behaviour_classes/base.py`, plus the `PACKAGE_QUERY` constant that was only used by `get_packages`. Removed the corresponding params from skill.yaml, agent.yaml, service.yaml and `models.py`:

- `service_registry_address_base` / `service_registry_address_celo` — only feeder for the deleted `get_service_registry_address`
- `olas_subgraph_url` — only feeder for the deleted `get_packages`

Test classes `TestGetPackages`, `TestGetPackagesJsonParseError`, and `TestAddressGetters.test_service_registry` are gone with them. Conftest fixtures stripped of the same params.

The `valory/genai` connection hash is also bumped to the post-review fixup version (`bafybeiacgeofhh...`) — the upstream PR (valory-xyz/genai#30) replaced the threading wrapper with the SDK's native `request_options.timeout` after the same reviewer flagged that the executor's `shutdown(wait=True)` defeated the deadline.

CI green locally — `tox -e py3.12-darwin` 1336 tests pass (down from 1349 because the orphaned tests are gone), `black-check / isort-check / flake8 / mypy / pylint / darglint / check-hash / check-packages / check-handlers / check-dependencies` all clean.

**Comment by DhairyaPatel7** [comment]
LGTM overall — six in-repo fixes look correct, dead-code deletion is thorough (the `re` import, the `MEMEOOORR_DESCRIPTION_PATTERN` constant, the `IPFS_ENDPOINT` constant, the `service_registry_l2` skill-yaml dependency, and the `ServiceRegistryContract` import all come out in lockstep), and the new tests exercise the partial-null and bad-body paths the audit called out.

Notes:

- `f"AgentDB request failed: {response.status_code} - {response.body!r}"` — `HttpMessage.body` is bytes, so `!r` produces a `b'…'`-prefixed repr in the operator-facing error. Strict improvement over `.text` (which doesn't exist on `HttpMessage` — the bug being fixed), but `body.decode("utf-8", errors="replace")` would be a cleaner message. Style-only.
- This PR's `chore(packages): bump genai + x402 connection hashes` carries the IPFS hashes from valory-xyz/genai#30. Given the open inline review on #30 the deadline implementation is very likely to change before merge — when it does, that commit will need to be rerolled against the post-merge hashes. The PR description already flags this, but worth holding this one until #30 lands.

**Comment by DIvyaNautiyal07** [comment]
Thanks for the read.

On the body decode nit: applied in `12ddcf6c`. The RuntimeError in `_request` now reads `body.decode('utf-8', errors='replace')` so the operator log shows `Server Error` instead of `b'Server Error'`. Left the parallel `!r` on the 200/201 non-JSON-body log untouched — bytes there may be HTML or control characters and the repr signals that explicitly. Existing test still passes since its regex matched on the substring rather than the repr framing.

On the hash bump waiting for #30: agreed. The current `chore(packages): bump genai + x402 connection hashes` commit on this branch holds the preview hashes I published from the fix/audit-70-genai-resilience branch. If review on #30 changes anything (still active — open thread on `GENAI_DIRECT_TIMEOUT_SECONDS`'s value), the hashes will move and that commit will need to be re-rolled. CI's `check-third-party-hashes` against the `valory-xyz/genai@v7.0.0` upstream pin is also blocking until this lands and a new tag is cut. We can hold this PR until #30 merges + a new tag is released, or split the hash bump out into a follow-up — either works.

**Review by DhairyaPatel7 — DISMISSED** [review]

File: packages/valory/connections/mirror_db/connection.py [hunk 1]
```diff
@@ -46,6 +46,9 @@
 DEFAULT_HEADERS = {"Content-Type": "application/json", "accept": "application/json"}
 
 
+RETRYABLE_5XX = frozenset({502, 503, 504})
```
Comment (LOCKhart07) [inline packages/valory/connections/mirror_db/connection.py:57]: Non-blocking design question: this `@retry_with_exponential_backoff()` is applied to `create_` (POST) at line 365, and to `update_` / `delete_` at lines 395/412, in addition to `read_`. Adding 502/504 to the retry set changes the duplicate-write profile here — a 502 (bad gateway) or 504 (gateway timeout) often means the upstream did receive the request, just that the gateway response was lost. On a non-idempotent POST that's a duplicate-create hazard. 429 and connection errors didn't have this property. 503 is generally safe (load-shedder, request rejected before processing).

If the mirror_db backend is idempotent on `create_` then this is fine — worth a one-line note in the commit body. Otherwise consider scoping the new 5xx retry to `read_` only, or to `503` only on the write methods.

File: packages/valory/skills/memeooorr_chained_abci/skill.yaml [hunk 2]
```diff
@@ -27,9 +27,9 @@ skills:
 - valory/reset_pause_abci:0.1.0:bafybeifdzwzz5byan3ytsi46n4n2svtqme6475hk5z77sopymyzqszcpoa
 - valory/transaction_settlement_abci:0.1.0:bafybeigiojqgbi54jhree3sqrzaudbmjxb7y2ete74bbliryi3t7ztwi5q
 - valory/termination_abci:0.1.0:bafybeiadbtzcvzkwbrosshlsubilj7mwzdqfni6dr75wsd3vawhslpwqeu
-- valory/memeooorr_abci:0.1.0:bafybeieoe2x2t663ued4fjfbzp7bqvniqgqhhqjecctk7lzvysvqdx5byu
+- valory/memeooorr_abci:0.1.0:bafybeidcy6hgn4kvzg6oj5tdnaxafj3hx4s5lnqrmz4mkl2fsihk4ldama
```
Comment (LOCKhart07) [inline packages/valory/skills/memeooorr_chained_abci/skill.yaml:30]: Cleanup gap — same class of orphan-params issue that drove `dc240fdc`. Lines 161–164 of this file still declare `service_registry_address_base`, `service_registry_address_celo`, and `olas_subgraph_url` under `models.params.args`, but the matching `_ensure(...)` calls were removed from `memeooorr_abci/models.py` in this PR. `memeooorr_chained_abci.models.Params` inherits from `MemeooorrParams` (`memeooorr_chained_abci/models.py:52`), so these are dead config — they sit in kwargs unused after `_ensure` no longer pops them.

The same cleanup was applied to `agents/memeooorr/aea-config.yaml`, `services/memeooorr/service.yaml`, `memeooorr_abci/skill.yaml`, the test conftest, and `test_models.py` in `dc240fdc`, but the chained skill's yaml was missed. Drop those three lines and re-lock.

File: packages/valory/connections/mirror_db/connection.py (hunk 1 above)
Comment (DIvyaNautiyal07) [reply packages/valory/connections/mirror_db/connection.py:57]: Substantive concern, picked up. Pushed `089d96bf` that scopes the 5xx retry to `read_` only.

`retry_with_exponential_backoff` and `_handle_retryable_exception` now both take a `retry_on_5xx: bool = False` flag. The default reverts writes to "retry on 429 + connection errors only" — the pre-P3-1 behaviour. `read_` is the only call site that opts in via `@retry_with_exponential_backoff(retry_on_5xx=True)`.

Tests in `089d96bf`:
- `test_transient_5xx_does_not_retry_by_default` (parametrized 502/503/504) confirms the default
- `test_transient_5xx_triggers_retry_when_opted_in` covers the read path
- `test_create_does_not_retry_on_transient_5xx` (parametrized) and `test_create_still_retries_on_429` exercise the full guarantee at the integration level — a `create_` call against a 502/503/504 raises immediately with `call_count == 1`, while a 429 still retries

Went with the third option from your comment ("scope writes to 503 only") taken to its conservative extreme — writes don't retry any 5xx — since 503 has the same "rejected vs processed" ambiguity 502/504 do once you stop assuming load-shedder semantics. Happy to relax that to `{503}` on writes if you'd rather keep some 5xx retry budget there.

File: packages/valory/skills/memeooorr_chained_abci/skill.yaml (hunk 2 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/memeooorr_chained_abci/skill.yaml:30]: Right, missed it. Dropped in `089d96bf`. The chained skill's yaml now matches the regular skill's — three lines (`service_registry_address_base`, `service_registry_address_celo`, `olas_subgraph_url`) gone, re-locked.

**Review by DhairyaPatel7 — DISMISSED** [review]

**Review by LOCKhart07 — APPROVED** [review]

**Review by DhairyaPatel7 — APPROVED** [review]

File: packages/valory/skills/memeooorr_abci/behaviour_classes/base.py [hunk 3]
```diff
@@ -825,7 +651,9 @@ def get_meme_coins_from_subgraph(self) -> Generator[None, None, Optional[List]]:
             )
             return []
 
-        items = (response_json.get("data") or {}).get("memeTokens", {}).get("items", [])
+        items = ((response_json.get("data") or {}).get("memeTokens") or {}).get(
```
Comment (OjusWiZard) [inline packages/valory/skills/memeooorr_abci/behaviour_classes/base.py:654]: While we're in this method — line 648 has `except (json.JSONDecodeError, Exception):  # pylint: disable=broad-except`. `Exception` subsumes `JSONDecodeError`, so the tuple is dead/misleading; the broad arm also swallows programmer-error `RuntimeError`/`AttributeError` at error-log level instead of letting them propagate.

Pre-existing (introduced 2026-03-17), and this PR actually removed two duplicate occurrences of the same pattern — leaving this last surviving instance. Worth narrowing in a follow-up to e.g. `except (json.JSONDecodeError, UnicodeDecodeError):` and dropping the pylint suppression. Not blocking.

File: packages/valory/skills/memeooorr_abci/tests/test_handlers.py [hunk 4]
```diff
@@ -1666,6 +1666,34 @@ def test_no_prompt_field(self) -> None:
...
+        handler = _make_http_handler()
+        handler.context.params.use_x402 = False
+        handler._handle_bad_request = MagicMock()
+        msg = _make_http_msg(method="post", body=b'{"prompt": "hi')
+        dialogue = _make_http_dialogue()
+
+        HttpHandler._handle_post_process_prompt(handler, msg, dialogue)
+
+        handler._handle_bad_request.assert_called_once()
+        args, _ = handler._handle_bad_request.call_args
```
Comment (OjusWiZard) [inline packages/valory/skills/memeooorr_abci/tests/test_handlers.py:1694]: Optional: these two tests differ only in the `body=` payload and the docstring, so a single `@pytest.mark.parametrize` over `[b"\xff\xfe", b'{"prompt": "hi']` would collapse them and make adding future bad-body cases trivial:

```python
@pytest.mark.parametrize("body", [b"\xff\xfe", b'{"prompt": "hi'])
def test_malformed_body_returns_bad_request(self, body: bytes) -> None:
    handler = _make_http_handler()
    handler.context.params.use_x402 = False
    handler._handle_bad_request = MagicMock()
    HttpHandler._handle_post_process_prompt(
        handler, _make_http_msg(method="post", body=body), _make_http_dialogue()
    )
    args, _ = handler._handle_bad_request.call_args
    assert "Malformed request body" in args[2]["error"]
```

Project convention is "three similar lines is fine" — only worth folding if more bad-body cases are anticipated. Take it or leave it.

**Review by OjusWiZard — APPROVED** [review]
Approving — the six in-repo fixes are correct and well-tested, and the dead-code sweep is clean (verified no surviving references to `ServiceRegistryContract`, `MEMEOOORR_DESCRIPTION_PATTERN`, `IPFS_ENDPOINT`, or the deleted helpers anywhere in the source tree).

The scoping in `089d96bf` is the right call: gating `RETRYABLE_5XX` behind `retry_on_5xx=True` and opting in only `read_` (idempotent GET) closes the duplicate-write hazard on 502/504, where the upstream may have processed the request before the gateway response was lost. Inline docstring on the constant calls this out clearly.

Two small advisory items inline — both pre-existing/optional, not blocking. One additional out-of-scope observation:

**Pre-existing, surfaced by the cleanup (follow-up only):** `EngageTwitterBehaviour.get_agent_handles` in `behaviour_classes/twitter.py:1527-1541` is annotated `Generator[None, None, List[AgentsFunAgent]]` but contains no `yield` — it `return`s a list directly. The call site at `twitter.py:655` uses `yield from self.get_agent_handles()`, which iterates the list (works), but the annotation no longer matches the runtime semantics. Not modified by this PR; flagging only because the PR's dead-code sweep brought attention to this area.

**Pre-merge note already discussed in the thread:** `chore(packages): bump genai + x402 connection hashes` (`ccaf2c07`) carries IPFS hashes from the unmerged `valory-xyz/genai#30`. Hold this PR until #30 lands and a new upstream tag is cut, or split the hash bump into a follow-up.

Nice work on the test quality — the `SimpleNamespace` swap for `test_non_success_status_raises_runtime_error` will catch any future re-introduction of `.text`, and the parametrized 502/503/504 sweep + the explicit `test_500_is_not_retryable` boundary test make the retry-set boundary essentially regression-proof.

File: packages/valory/skills/memeooorr_abci/tests/test_handlers.py (hunk 4 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/memeooorr_abci/tests/test_handlers.py:1694]: will address in follow-up

File: packages/valory/skills/memeooorr_abci/behaviour_classes/base.py (hunk 3 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/memeooorr_abci/behaviour_classes/base.py:654]: will address in follow-up

Comment (DIvyaNautiyal07) [reply packages/valory/skills/memeooorr_abci/behaviour_classes/base.py:654]: Picked up in [9d34d976](https://github.com/valory-xyz/meme-ooorr/commit/9d34d976) on the FSM branch (#71): narrowed to `(json.JSONDecodeError, UnicodeDecodeError)` and dropped the pylint suppression. Added `test_undecodable_bytes_body_returns_empty` to exercise the `UnicodeDecodeError` arm explicitly — the existing `test_non_json_body_returns_empty` already covered `JSONDecodeError`.

File: packages/valory/skills/memeooorr_abci/tests/test_handlers.py (hunk 4 above)
Comment (DIvyaNautiyal07) [reply packages/valory/skills/memeooorr_abci/tests/test_handlers.py:1694]: Folded in [dfc5441a](https://github.com/valory-xyz/meme-ooorr/commit/dfc5441a) on the FSM branch (#71). Collapsed the two methods under `@pytest.mark.parametrize` over `[(b"\xff\xfe", "non-utf8 bytes"), (b'{"prompt": "hi', "truncated json")]`. Adding a future bad-body case is now one tuple instead of a copy-pasted method.
